src/sidebar/components/tag-editor.js

```diff
@@ -1,35 +1,197 @@
+    onEditTags({ tags: tagList });
   };
 
-  this.autocomplete = function(query) {
-    return Promise.resolve(tags.filter(query));
+  /**
+   * Remove a tag from this annotation.
+   *
+   * @param {string} tag
+   */
+  const removeTag = tag => {
+    const newTagList = [...tagList]; // make a copy
+    const index = newTagList.indexOf(tag);
+    newTagList.splice(index, 1);
+    updateTags(newTagList);
   };
 
-  this.$onChanges = function(changes) {
-    if (changes.tags) {
-      this.tagList = changes.tags.currentValue.map(function(tag) {
-        return { text: tag };
+  /**
+   * Adds a tag equal to the value of the input field, and
+   * then clears out the suggestions list and the input field.
+   */
+  const addTag = () => {
+    const value = inputEl.current.value.trim();
+    if (value.length === 0) {
+      // don't add an empty tag
+      return;
+    }
+    if (tagList.indexOf(value) >= 0) {
+      // don't add duplicate tag
+      return;
+    }
+
+    updateTags([...tagList, value]);
+    setSuggestions([]);
+
+    // clear the input field and maintain focus
+    inputEl.current.value = '';
+    inputEl.current.focus();
+  };
+
+  /**
+   *  If the user pressed enter or comma while focused on
+   *  the <input>, then add a new tag.
+   */
+  const handleKeyPress = e => {
+    if (e.key === 'Enter' || e.key === ',') {
+      addTag();
+      // preventDefault stops the delimiter from being
+      // added to the input field
+      e.preventDefault();
+    }
+  };
+  /**
+   *  As the user types, change the suggestions query
+   *  based on the value of inputEl.
+   */
+  const handleKeyUp = e => {
+    // Helper to remove any repeated suggestions
+    // that are already tags.
+    const removeDuplicates = suggestions => {
+      const suggestionsSet = [];
+      suggestions.forEach(suggestion => {
+        if (tagList.indexOf(suggestion) < 0) {
+          suggestionsSet.push(suggestion);
+        }
       });
+      return suggestionsSet;
+    };
+
+    if (e.key && e.key !== 'Enter') {
+      // A user typed a letter and we need to filter the results
+      // to update the <dataset> suggestions
+      const suggestions = removeDuplicates(
+        tagsService.filter(inputEl.current.value)
+      );
+      setSuggestions(suggestions);
+    } else if (e.type === 'keyup') {
+      // Assume the user selected an option from <datalist>
+      // either by clicking on it with the mouse or pressing
+      // the 'Enter' key.
+      addTag();
     }
   };
+
+  const suggestionsList = (() => {
+    const suggestionElements = suggestions.map((suggestion, index) => (
+      <option key={index} value={suggestion} />
+    ));
+
+    // Only return a datalist if there is something to render in the list. This
+    // avoid a bug in Firefox where the datalist rendered does not correspond to the
+    // actual list of <option>s in the dom. Wiping out the entire <datalist> tag seems
+    // to force Firefox's hand.
```

Comment:
Is there a link we could put here to information about the relevant bug in FF?

Reply:
Don't need this anymore after not changing datalist options around.